lib/base/journaldlogger.cpp

```diff
@@ -0,0 +1,81 @@
+/* Icinga 2 | (c) 2021 Icinga GmbH | GPLv2+ */
+
+#include "base/i2-base.hpp"
+#if !defined(_WIN32) && defined(HAVE_SYSTEMD)
+#include "base/journaldlogger.hpp"
+#include "base/journaldlogger-ti.cpp"
+#include "base/configtype.hpp"
+#include "base/statsfunction.hpp"
+#include "base/sysloglogger.hpp"
+#include <sys/uio.h>
+#include <systemd/sd-journal.h>
+
+using namespace icinga;
+
+REGISTER_TYPE(JournaldLogger);
+
+REGISTER_STATSFUNCTION(JournaldLogger, &JournaldLogger::StatsFunc);
+
+void JournaldLogger::StatsFunc(const Dictionary::Ptr& status, const Array::Ptr&)
+{
+	DictionaryData nodes;
+
+	for (const JournaldLogger::Ptr& journaldlogger : ConfigType::GetObjectsByType<JournaldLogger>()) {
+		nodes.emplace_back(journaldlogger->GetName(), 1); //add more stats
+	}
+
+	status->Set("journaldlogger", new Dictionary(std::move(nodes)));
+}
+
+void JournaldLogger::OnConfigLoaded()
+{
+	ObjectImpl<JournaldLogger>::OnConfigLoaded();
+	m_Facility = SyslogHelper::FacilityToNumber(GetFacility());
+}
+
+void JournaldLogger::ValidateFacility(const Lazy<String>& lvalue, const ValidationUtils& utils)
+{
+	ObjectImpl<JournaldLogger>::ValidateFacility(lvalue, utils);
+	if (!SyslogHelper::ValidateFacility(lvalue()))
+		BOOST_THROW_EXCEPTION(ValidationError(this, { "facility" }, "Invalid facility specified."));
+}
+
+/**
+ * Processes a log entry and outputs it to journald.
+ *
+ * @param entry The log entry.
+ */
+void JournaldLogger::ProcessLogEntry(const LogEntry& entry)
+{
+	std::vector<String> sdFields {
+		String("MESSAGE=") + entry.Message.GetData(),
+		String("PRIORITY=") + Value(SyslogHelper::SeverityToNumber(entry.Severity)),
+		String("SYSLOG_FACILITY=") + Value(m_Facility),
+		String("ICINGA2_FACILITY=") + entry.Facility,
+	};
+	const String identifier = GetIdentifier();
+	if (!identifier.IsEmpty()) {
+		sdFields.push_back(String("SYSLOG_IDENTIFIER=" + identifier));
+	}
```

Comment:
This could be split into a static and a dynamic part, i.e. initialize the values for `SYSLOG_FACILITY` and `SYSLOG_IDENTIFIER` once in `OnConfigLoaded()` to avoid allocating and concatenating these every single time. `SystemdJournalSend()` could then merge both of these into the `iovec[]`.

Reply:
Good idea, tried sort of that.